src/mdformat/_api.py

```diff
@@ -68,7 +68,11 @@ def file(
     )
     newline = (
         "\r\n"
-        if options.get("end_of_line", DEFAULT_OPTS["end_of_line"]) == "crlf"
+        if (
+            options.get("end_of_line", DEFAULT_OPTS["end_of_line"]) == "crlf"
+            or options.get("end_of_line", DEFAULT_OPTS["end_of_line"]) == "crlf"
+            and original_md[original_md.find("\n") - 1] == "\r"
+        )
         else "\n"
     )
```

Comment:
I think the conditional expression is getting too long to be readable here. I suggest replacing this with
```python
eol_setting = options.get("end_of_line", DEFAULT_OPTS["end_of_line"])
if eol_setting == "keep":
    first_eol = re.search(r"\r\n|\r|\n", original_str)
    newline = "\r\n" if first_eol and first_eol.group() == "\r\n" else "\n"
elif eol_setting == "crlf":
    newline = "\r\n"
else:
    newline = "\n"
```

And maybe also move this to a function (that takes `eol_setting` string as input) and share the code with `_cli.py`

Reply:
yea, externalising this makes sense. I also have to think about this again now that we defnetly want to upport `\r`.

Reply:
also sidenote: it would probably be sensible to rewrite this at some point so that the cli uses the api code and not having two version of nearly identical code. But thats a lot of work for some other time

Reply:
:+1: The code block I provided should work fine with `\r`. It'll convert them to `\n` as per the specification in https://github.com/executablebooks/mdformat/issues/336#issuecomment-1224883207

Reply:
I also just read it completely and realsied you did more than just untangle the `if`

I also tested `first_eol = re.search(r"\r\n|\r|\n", original_str)` and found something wird with it, but can't remember what. I will retest and come back to this